component/discovery/kubelet/kubelet.go

```diff
@@ -0,0 +1,268 @@
+// Package kubelet implements a discovery.kubelet component.
+package kubelet
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/grafana/agent/component"
+	"github.com/grafana/agent/component/common/config"
+	"github.com/grafana/agent/component/discovery"
+	discoveryk8s "github.com/grafana/agent/component/discovery/kubernetes"
+	commonConfig "github.com/prometheus/common/config"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/discovery/refresh"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+	"github.com/prometheus/prometheus/util/strutil"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	defaultKubeletRefreshInterval = time.Minute
+
+	metaLabelPrefix               = model.MetaLabelPrefix + "kubernetes_"
+	namespaceLabel                = metaLabelPrefix + "namespace"
+	nodeNameLabel                 = metaLabelPrefix + "node_name"
+	presentValue                  = model.LabelValue("true")
+	podNameLabel                  = metaLabelPrefix + "pod_name"
+	podIPLabel                    = metaLabelPrefix + "pod_ip"
+	podContainerNameLabel         = metaLabelPrefix + "pod_container_name"
+	podContainerImageLabel        = metaLabelPrefix + "pod_container_image"
+	podContainerPortNameLabel     = metaLabelPrefix + "pod_container_port_name"
+	podContainerPortNumberLabel   = metaLabelPrefix + "pod_container_port_number"
+	podContainerPortProtocolLabel = metaLabelPrefix + "pod_container_port_protocol"
+	podContainerIsInit            = metaLabelPrefix + "pod_container_init"
+	podReadyLabel                 = metaLabelPrefix + "pod_ready"
+	podPhaseLabel                 = metaLabelPrefix + "pod_phase"
+	podLabelPrefix                = metaLabelPrefix + "pod_label_"
+	podLabelPresentPrefix         = metaLabelPrefix + "pod_labelpresent_"
+	podAnnotationPrefix           = metaLabelPrefix + "pod_annotation_"
+	podAnnotationPresentPrefix    = metaLabelPrefix + "pod_annotationpresent_"
+	podNodeNameLabel              = metaLabelPrefix + "pod_node_name"
+	podHostIPLabel                = metaLabelPrefix + "pod_host_ip"
+	podUID                        = metaLabelPrefix + "pod_uid"
+	podControllerKind             = metaLabelPrefix + "pod_controller_kind"
+	podControllerName             = metaLabelPrefix + "pod_controller_name"
+)
+
+func init() {
+	component.Register(component.Registration{
+		Name:    "discovery.kubelet",
+		Args:    Arguments{},
+		Exports: discovery.Exports{},
+		Build: func(opts component.Options, args component.Arguments) (component.Component, error) {
+			return New(opts, args.(Arguments))
+		},
+	})
+}
+
+// Arguments configures the discovery.kubernetes component.
+type Arguments struct {
+	KubeletURL       config.URL                        `river:"kubelet_url,attr"`
```

Comment:
I had some open questions over in https://github.com/grafana/agent/issues/4233#issuecomment-1603281700 where I'm still not sure of the answers:

> * How can a container get access to the Kubelet API? I know you can proxy requests to it from the API server, but that's what we're trying to avoid. Is there a way containers can talk directly to their hosting Kubelets?
> * When talking directly to a Kubelet, it's not clear what authentication we would need to use.

This is stuff I think we should probably end up documenting, so can you help me understand these points? 

How would someone set this URL correctly in Kubernetes? What would they need to do to make the Kubelet accessible (if anything?)

And where would the authentication come from?

Reply:
Hey @rfratto thanks for taking a look!
- This is going to depend on how you have networking setup between hosts and pods, but the kubelet has an `--address` flag to bind it to an interface and then exposes a REST https server on 10250 by default (the port can be changed with `--port`) so if your agent pod can route to `<host-ip>:10250` and you bind the kubelet to `0.0.0.0` this will work, or if you have your grafana-agent daemonset host networked this also works. 
The `/pods` endpoint on the kubelet I'm using in this PR is essentially what the API server proxies too, this PR bypasses the API server and hits the kubelets directly. Essentially what we need to call out in the docs is that the grafana-agent pod must be able to reach the kubelet port for this to work.
- Authentication wise you can use a normal bearer token from a `serviceaccount` you just need `get` permissions on `nodes/proxy` resource in the grafana-agent cluster role, client certs should also work. Theres an also an option to enable an unauthed read only port on the kubelet, but it's not recommended.

Reply:
Thanks for helping me understand! 

> Authentication wise you can use a normal bearer token from a serviceaccount you just need get permissions on nodes/proxy resource in the grafana-agent cluster role, client certs should also work. Theres an also an option to enable an unauthed read only port on the kubelet, but it's not recommended.

Oh interesting! I was aware of the nodes/proxy resource but I didn't realize that it could also be used to talk to the nodes directly. Good to know 👍